util/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -100,17 +113,28 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     OutlierDetectionLoadBalancerConfig config
         = (OutlierDetectionLoadBalancerConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
 
-    // The map should only retain entries for addresses in this latest update.
-    ArrayList<SocketAddress> addresses = new ArrayList<>();
+    // The map should only retain entries for endpoints in this latest update.
+    Set<Set<SocketAddress>> endpoints = new HashSet<>();
+    Map<SocketAddress, Set<SocketAddress>> addressEndpointMap = new HashMap<>();
     for (EquivalentAddressGroup addressGroup : resolvedAddresses.getAddresses()) {
-      addresses.addAll(addressGroup.getAddresses());
+      Set<SocketAddress> endpoint = ImmutableSet.copyOf(addressGroup.getAddresses());
+      endpoints.add(endpoint);
+      for (SocketAddress address : addressGroup.getAddresses()) {
+        addressEndpointMap.putIfAbsent(address, endpoint);
```

Comment:
Why are we using `putIfAbsent()` needs to be used instead of just `put()`? Using `put()` would always produce the same map as `putIfAbsent()`, right? Is this for optimization (it's probably not helping)?

Reply:
yes, a `put` is good and cleaner.